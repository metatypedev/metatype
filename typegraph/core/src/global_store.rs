// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::{self, Result};
use crate::runtimes::{DenoMaterializer, Materializer, MaterializerDenoModule, Runtime};
use crate::types::{Struct, Type, TypeFun, TypeId, WrapperTypeData};
use crate::wit::core::{Policy as CorePolicy, PolicyId, RuntimeId};
use crate::wit::runtimes::{Effect, MaterializerDenoPredefined, MaterializerId};
use graphql_parser::parse_query;
use indexmap::IndexMap;
use std::rc::Rc;
use std::{cell::RefCell, collections::HashMap};

pub type Policy = Rc<CorePolicy>;

/// As all the store entries are append only, we can set a restore point
/// to reset it to a previous state.
/// This is useful to remove the entities that were added in a typegraph scope
/// instead of the global scope.
///
/// The state is saved as item count for each entry.
///
/// With this feature, we can reuse a type name in a typegraph definition module,
/// within different typegraph contexts.
#[derive(Default, Debug)]
pub struct SavedState {
    types: usize,
    type_names: usize,
    runtimes: usize,
    materializers: usize,
    policies: usize,
}

#[derive(Default)]
pub struct Store {
    pub types: Vec<Type>,
    pub type_by_names: IndexMap<String, TypeId>,

    pub runtimes: Vec<Runtime>,
    pub materializers: Vec<Materializer>,
    pub policies: Vec<Policy>,

    deno_runtime: RuntimeId,
    predefined_deno_functions: HashMap<String, MaterializerId>,
    deno_modules: HashMap<String, MaterializerId>,

    prisma_migration_runtime: RuntimeId,
    typegate_runtime: RuntimeId,
    typegraph_runtime: RuntimeId,
    graphql_endpoints: Vec<String>,
}

impl Store {
    fn new() -> Self {
        Self {
            runtimes: vec![
                Runtime::Deno,
                Runtime::PrismaMigration,
                Runtime::Typegate,
                Runtime::Typegraph,
            ],
            deno_runtime: 0,
            prisma_migration_runtime: 1,
            typegate_runtime: 2,
            typegraph_runtime: 3,
            ..Default::default()
        }
    }
}

const PREDEFINED_DENO_FUNCTIONS: &[&str] = &["identity", "true"];

thread_local! {
    pub static STORE: RefCell<Store> = RefCell::new(Store::new());
}

fn with_store<T, F: FnOnce(&Store) -> T>(f: F) -> T {
    STORE.with(|s| f(&s.borrow()))
}

fn with_store_mut<T, F: FnOnce(&mut Store) -> T>(f: F) -> T {
    STORE.with(|s| f(&mut s.borrow_mut()))
}

#[cfg(test)]
impl Store {
    pub fn reset() {
        let _ = crate::typegraph::finalize();
        with_store_mut(|s| *s = Store::new());
    }
}

impl Store {
    pub fn save() -> SavedState {
        with_store(|s| SavedState {
            types: s.types.len(),
            type_names: s.type_by_names.len(),
            runtimes: s.runtimes.len(),
            materializers: s.materializers.len(),
            policies: s.policies.len(),
        })
    }

    pub fn restore(saved_state: SavedState) {
        with_store_mut(|s| {
            s.types.truncate(saved_state.types);
            s.type_by_names.truncate(saved_state.type_names);
            s.runtimes.truncate(saved_state.runtimes);
            s.materializers.truncate(saved_state.materializers);
            s.policies.truncate(saved_state.policies);
        })
    }

    pub fn get_type_by_name(name: &str) -> Option<TypeId> {
        with_store(|s| s.type_by_names.get(name).copied())
    }

    pub fn register_type(build: impl FnOnce(TypeId) -> Type) -> Result<TypeId> {
        // this works since the store is thread local
        let id = with_store(|s| s.types.len()) as u32;
        let typ = build(id.into());
        if let Some(name) = typ.get_base().and_then(|b| b.name.clone()) {
            Self::register_type_name(name, id.into())?;
        }
        with_store_mut(move |s| -> Result<()> {
            s.types.push(typ);
            Ok(())
        })?;
        Ok(id.into())
    }

    pub fn register_type_name(name: impl Into<String>, id: TypeId) -> Result<()> {
        let name = name.into();
        with_store_mut(move |s| -> Result<()> {
            if s.type_by_names.contains_key(&name) {
                return Err(format!("type with name {:?} already exists", name).into());
            }
            s.type_by_names.insert(name, id);
            Ok(())
        })
    }

    pub fn get_type_by_path(struct_id: TypeId, path: &[String]) -> Result<(Type, TypeId)> {
        let mut ret = (struct_id.as_type()?, struct_id);

        let mut curr_path = vec![];
        for (pos, chunk) in path.iter().enumerate() {
            let unwrapped_id = ret.1.resolve_wrapper()?;
            // let unwrapped_id = self.resolve_wrapper(ret.1)?;
            match unwrapped_id.as_type()? {
                Type::Struct(t) => {
                    let result = t.data.props.iter().find(|(k, _)| k.eq(chunk));
                    curr_path.push(chunk.clone());
                    ret = match result {
                        Some((_, id)) => (TypeId(*id).as_type()?, id.to_owned().into()),
                        None => {
                            return Err(errors::invalid_path(
                                pos,
                                path,
                                &t.data
                                    .props
                                    .iter()
                                    .map(|v| format!("{:?}", v.0.clone()))
                                    .collect::<Vec<String>>(),
                            ));
                        }
                    };
                }
                _ => return Err(errors::expect_object_at_path(&curr_path)),
            }
        }

        Ok(ret)
    }

    pub fn register_runtime(rt: Runtime) -> RuntimeId {
        with_store_mut(|s| {
            let id = s.runtimes.len() as u32;
            s.runtimes.push(rt);
            id
        })
    }

    pub fn get_runtime(id: RuntimeId) -> Result<Runtime> {
        with_store(|s| {
            s.runtimes
                .get(id as usize)
                .cloned()
                .ok_or_else(|| errors::object_not_found("runtime", id))
        })
    }

    pub fn get_deno_runtime() -> RuntimeId {
        with_store(|s| s.deno_runtime)
    }

    pub fn get_prisma_migration_runtime() -> RuntimeId {
        with_store(|s| s.prisma_migration_runtime)
    }

    pub fn get_typegate_runtime() -> RuntimeId {
        with_store(|s| s.typegate_runtime)
    }

    pub fn get_typegraph_runtime() -> RuntimeId {
        with_store(|s| s.typegraph_runtime)
    }

    pub fn register_materializer(mat: Materializer) -> MaterializerId {
        with_store_mut(|s| {
            let id = s.materializers.len() as u32;
            s.materializers.push(mat);
            id
        })
    }

    pub fn get_materializer(id: MaterializerId) -> Result<Materializer> {
        with_store(|s| {
            s.materializers
                .get(id as usize)
                .cloned()
                .ok_or_else(|| errors::object_not_found("materializer", id))
        })
    }

    pub fn register_policy(policy: Policy) -> Result<PolicyId> {
        with_store_mut(|s| {
            let id = s.policies.len() as u32;
            if s.policies.iter().any(|p| p.name == policy.name) {
                Err(errors::duplicate_policy_name(&policy.name))
            } else {
                s.policies.push(policy);
                Ok(id)
            }
        })
    }

    pub fn get_policy(id: PolicyId) -> Result<Policy> {
        with_store(|s| {
            s.policies
                .get(id as usize)
                .cloned()
                .ok_or_else(|| errors::object_not_found("policy", id))
        })
    }

    pub fn get_predefined_deno_function(name: String) -> Result<MaterializerId> {
        if let Some(mat) = with_store(|s| s.predefined_deno_functions.get(&name).cloned()) {
            Ok(mat)
        } else if !PREDEFINED_DENO_FUNCTIONS.iter().any(|n| n == &name) {
            Err(errors::unknown_predefined_function(&name, "deno"))
        } else {
            let runtime_id = Store::get_deno_runtime();
            let mat = Store::register_materializer(Materializer {
                runtime_id,
                effect: Effect::Read,
                data: Rc::new(DenoMaterializer::Predefined(MaterializerDenoPredefined {
                    name: name.clone(),
                }))
                .into(),
            });
            with_store_mut(|s| {
                s.predefined_deno_functions.insert(name, mat);
            });
            Ok(mat)
        }
    }

    pub fn get_deno_module(file: String) -> MaterializerId {
        if let Some(mat) = with_store(|s| s.deno_modules.get(&file).cloned()) {
            mat
        } else {
            let runtime_id = Store::get_deno_runtime();
            let mat = Store::register_materializer(Materializer {
                runtime_id,
                effect: Effect::Read, // N/A
                data: Rc::new(DenoMaterializer::Module(MaterializerDenoModule {
                    file: file.clone(),
                }))
                .into(),
            });
            with_store_mut(|s| s.deno_modules.insert(file, mat));
            mat
        }
    }

    pub fn add_graphql_endpoint(graphql: String) -> Result<()> {
        with_store_mut(|s| {
            let ast = parse_query::<&str>(&graphql).map_err(|e| e.to_string())?;
            let endpoints = ast
                .definitions
                .into_iter()
                .map(|op| {
                    format!("{}", op)
                        .split_whitespace()
                        .collect::<Vec<_>>()
                        .join(" ")
                })
                .collect::<Vec<_>>();

            s.graphql_endpoints.extend(endpoints);
            Ok(())
        })
    }

    pub fn get_graphql_endpoints() -> Vec<String> {
        with_store(|s| s.graphql_endpoints.clone())
    }
}

impl TypeId {
    pub fn as_type(&self) -> Result<Type> {
        with_store(|s| {
            s.types
                .get(self.0 as usize)
                .cloned()
                .ok_or_else(|| errors::object_not_found("type", self.0))
        })
    }

    pub fn as_struct(&self) -> Result<Rc<Struct>> {
        match self.as_type()? {
            Type::Struct(s) => Ok(s),
            Type::Proxy(inner) => inner.data.try_resolve()?.as_struct(),
            _ => Err(errors::invalid_type("Struct", &self.repr()?)),
        }
    }

    pub fn is_func(&self) -> Result<bool> {
        Ok(matches!(self.as_type()?, Type::Func(_)))
    }

    pub fn resolve_quant(&self) -> Result<TypeId> {
        let type_id = *self;
        match type_id.as_type()? {
            Type::Array(a) => Ok(a.data.of.into()),
            Type::Optional(o) => Ok(o.data.of.into()),
            _ => Ok(type_id),
        }
    }

    /// unwrap type id inside array, optional, or WithInjection
    pub fn resolve_wrapper(&self) -> Result<TypeId> {
        let mut id = self.resolve_proxy()?;
        loop {
            let tpe = id.as_type()?;
            let new_id = match tpe {
                Type::Array(t) => t.data.of.into(),
                Type::Optional(t) => t.data.of.into(),
                Type::WithInjection(t) => t.data.tpe.into(),
                Type::Proxy(t) => t.id.resolve_proxy()?,
                _ => id,
            };
            if id == new_id {
                break;
            }
            id = new_id;
        }
        Ok(id)
    }
}
