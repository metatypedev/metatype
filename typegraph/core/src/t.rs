// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::errors::Result;
use crate::types::TypeId;
use crate::wit::core::{
    Core, TypeArray, TypeBase, TypeEither, TypeFloat, TypeFunc, TypeInteger, TypeOptional,
    TypeProxy, TypeString, TypeStruct, TypeUnion,
};

pub trait TypeBuilder {
    fn build(&mut self) -> Result<TypeId>;
}

pub trait ConcreteTypeBuilder: TypeBuilder {
    fn base_mut(&mut self) -> &mut TypeBase;

    fn named(&mut self, name: impl Into<String>) -> &mut Self {
        self.base_mut().name = Some(name.into());
        self
    }

    fn as_id(&mut self, as_id: bool) -> &mut Self {
        self.base_mut().as_id = as_id;
        self
    }

    fn config(&mut self, key: impl Into<String>, value: impl Into<String>) -> &mut Self {
        let runtime_config = &mut self.base_mut().runtime_config;
        if runtime_config.is_none() {
            *runtime_config = Some(Default::default());
        }
        runtime_config
            .as_mut()
            .unwrap()
            .push((key.into(), value.into()));
        self
    }
}

#[derive(Default)]
pub struct BooleanBuilder {
    base: TypeBase,
}

pub fn boolean() -> BooleanBuilder {
    Default::default()
}

#[derive(Default)]
pub struct IntegerBuilder {
    base: TypeBase,
    data: TypeInteger,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeInteger {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            exclusive_minimum: None,
            exclusive_maximum: None,
            multiple_of: None,
            enumeration: None,
        }
    }
}

impl IntegerBuilder {
    #[allow(dead_code)]
    pub fn min(mut self, min: i32) -> Self {
        self.data.min = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn max(mut self, max: i32) -> Self {
        self.data.max = Some(max);
        self
    }

    #[allow(dead_code)]
    pub fn x_min(mut self, min: i32) -> Self {
        self.data.exclusive_minimum = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn x_max(mut self, max: i32) -> Self {
        self.data.exclusive_maximum = Some(max);
        self
    }
}

pub fn integer() -> IntegerBuilder {
    Default::default()
}

#[derive(Default)]
pub struct FloatBuilder {
    base: TypeBase,
    data: TypeFloat,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeFloat {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            exclusive_minimum: None,
            exclusive_maximum: None,
            multiple_of: None,
            enumeration: None,
        }
    }
}

impl FloatBuilder {
    #[allow(dead_code)]
    pub fn min(mut self, min: f64) -> Self {
        self.data.min = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn max(mut self, max: f64) -> Self {
        self.data.max = Some(max);
        self
    }

    #[allow(dead_code)]
    pub fn x_min(mut self, min: f64) -> Self {
        self.data.exclusive_minimum = Some(min);
        self
    }

    #[allow(dead_code)]
    pub fn x_max(mut self, max: f64) -> Self {
        self.data.exclusive_maximum = Some(max);
        self
    }
}

pub fn float() -> FloatBuilder {
    Default::default()
}

#[derive(Default)]
pub struct StringBuilder {
    base: TypeBase,
    data: TypeString,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeString {
    fn default() -> Self {
        Self {
            min: None,
            max: None,
            format: None,
            pattern: None,
            enumeration: None,
        }
    }
}

pub fn string() -> StringBuilder {
    Default::default()
}

impl StringBuilder {
    #[allow(dead_code)]
    pub fn format(&mut self, format: impl Into<String>) -> &mut Self {
        self.data.format = Some(format.into());
        self
    }

    pub fn enum_(&mut self, values: Vec<String>) -> &mut Self {
        self.data.enumeration = Some(
            values
                .into_iter()
                .map(|v| serde_json::to_string(&serde_json::Value::String(v)).unwrap())
                .collect(),
        );
        self
    }
}

#[derive(Default)]
pub struct OptionalBuilder {
    base: TypeBase,
    data: TypeOptional,
}

impl Default for TypeOptional {
    fn default() -> Self {
        Self {
            of: u32::max_value(),
            default_item: None,
        }
    }
}

pub fn optional(ty: TypeId) -> OptionalBuilder {
    OptionalBuilder {
        base: TypeBase::default(),
        data: TypeOptional {
            of: ty.into(),
            default_item: None,
        },
    }
}

#[derive(Default)]
pub struct ArrayBuilder {
    base: TypeBase,
    data: TypeArray,
}

impl Default for TypeArray {
    fn default() -> Self {
        Self {
            of: u32::max_value(),
            min: None,
            max: None,
            unique_items: None,
        }
    }
}

pub fn array(ty: TypeId) -> ArrayBuilder {
    ArrayBuilder {
        base: TypeBase::default(),
        data: TypeArray {
            of: ty.into(),
            ..Default::default()
        },
    }
}

#[derive(Default)]
pub struct UnionBuilder {
    base: TypeBase,
    data: TypeUnion,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeUnion {
    fn default() -> Self {
        Self {
            variants: Default::default(),
        }
    }
}

pub fn union(variants: impl IntoIterator<Item = TypeId>) -> UnionBuilder {
    UnionBuilder {
        data: TypeUnion {
            variants: variants.into_iter().map(|tid| tid.0).collect(),
        },
        ..Default::default()
    }
}

#[derive(Default)]
pub struct EitherBuilder {
    base: TypeBase,
    data: TypeEither,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeEither {
    fn default() -> Self {
        Self {
            variants: Default::default(),
        }
    }
}

pub fn either(variants: impl IntoIterator<Item = TypeId>) -> EitherBuilder {
    EitherBuilder {
        data: TypeEither {
            variants: variants.into_iter().map(|tid| tid.0).collect(),
        },
        ..Default::default()
    }
}

#[derive(Default)]
pub struct StructBuilder {
    base: TypeBase,
    data: TypeStruct,
}

#[allow(clippy::derivable_impls)]
impl Default for TypeStruct {
    fn default() -> Self {
        Self {
            props: Vec::new(),
            additional_props: false,
            min: None,
            max: None,
        }
    }
}

pub fn struct_() -> StructBuilder {
    Default::default()
}

pub fn struct_from(props: impl Iterator<Item = (String, TypeId)>) -> StructBuilder {
    StructBuilder {
        data: TypeStruct {
            props: props.map(|(k, ty)| (k, ty.into())).collect(),
            ..Default::default()
        },
        ..Default::default()
    }
}

pub fn struct_extends(ty: TypeId) -> Result<StructBuilder> {
    Ok(StructBuilder {
        data: TypeStruct {
            props: ty.as_struct().map(|typ| typ.data.props.clone())?,
            ..Default::default()
        },
        ..Default::default()
    })
}

impl StructBuilder {
    pub fn prop(&mut self, name: impl Into<String>, ty: TypeId) -> &mut Self {
        self.data.props.push((name.into(), ty.into()));
        self
    }

    #[allow(dead_code)]
    pub fn props(&mut self, props: impl IntoIterator<Item = (String, TypeId)>) {
        self.data
            .props
            .extend(props.into_iter().map(|(name, ty)| (name, ty.into())));
    }

    pub fn min(&mut self, min: u32) -> &mut Self {
        self.data.min = Some(min);
        self
    }

    pub fn max(&mut self, max: u32) -> &mut Self {
        self.data.max = Some(max);
        self
    }
}

#[derive(Default)]
pub struct FuncBuilder {
    #[allow(dead_code)]
    base: TypeBase,
    data: TypeFunc,
}

impl Default for TypeFunc {
    fn default() -> Self {
        Self {
            inp: u32::max_value(),
            out: u32::max_value(),
            mat: u32::max_value(),
        }
    }
}

#[allow(dead_code)]
pub fn func(inp: TypeId, out: TypeId, mat: u32) -> Result<TypeId> {
    FuncBuilder {
        data: TypeFunc {
            inp: inp.into(),
            out: out.into(),
            mat,
        },
        ..Default::default()
    }
    .build()
}

#[derive(Default)]
pub struct ProxyBuilder {
    data: TypeProxy,
}

impl Default for TypeProxy {
    fn default() -> Self {
        Self {
            name: "".to_string(),
            extras: vec![],
        }
    }
}

impl ProxyBuilder {
    /// Adds extra data entry in the proxy
    pub fn set(&mut self, key: impl Into<String>, value: impl Into<String>) -> &mut Self {
        self.data.extras.push((key.into(), value.into()));
        self
    }
}

pub fn proxy(name: impl Into<String>) -> ProxyBuilder {
    ProxyBuilder {
        data: TypeProxy {
            name: name.into(),
            ..Default::default()
        },
    }
}

macro_rules! impl_type_builder {
    ( $ty:ty, $build:ident ) => {
        impl TypeBuilder for $ty {
            fn build(&mut self) -> Result<TypeId> {
                let builder = std::mem::take(self);
                Ok($crate::Lib::$build(builder.data, builder.base)?.into())
            }
        }

        impl ConcreteTypeBuilder for $ty {
            fn base_mut(&mut self) -> &mut TypeBase {
                &mut self.base
            }
        }
    };

    ( $ty:ty, $build:ident, true ) => {
        impl TypeBuilder for $ty {
            fn build(&mut self) -> Result<TypeId> {
                let builder = std::mem::take(self);
                Ok($crate::Lib::$build(builder.data)?.into())
            }
        }
    };
}
impl TypeBuilder for BooleanBuilder {
    fn build(&mut self) -> Result<TypeId> {
        let builder = std::mem::take(self);
        Ok(crate::Lib::booleanb(builder.base)?.into())
    }
}

impl ConcreteTypeBuilder for BooleanBuilder {
    fn base_mut(&mut self) -> &mut TypeBase {
        &mut self.base
    }
}

impl_type_builder!(IntegerBuilder, integerb);
impl_type_builder!(FloatBuilder, floatb);
impl_type_builder!(OptionalBuilder, optionalb);
impl_type_builder!(StringBuilder, stringb);
impl_type_builder!(ArrayBuilder, arrayb);
impl_type_builder!(UnionBuilder, unionb);
impl_type_builder!(EitherBuilder, eitherb);
impl_type_builder!(StructBuilder, structb);
impl_type_builder!(FuncBuilder, funcb, true);
impl_type_builder!(ProxyBuilder, proxyb, true);
