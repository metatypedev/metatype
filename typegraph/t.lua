wrk.method = "POST"
wrk.body = '{"operationName":null,"variables":{},"query":"{add(a: 2, b: 4)currentToken indexPlayground(a: 6)}"}'
wrk.headers["Content-Type"] = "application/json"
wrk.headers["token"] = "test2"

