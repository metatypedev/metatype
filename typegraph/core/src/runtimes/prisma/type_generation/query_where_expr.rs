// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use crate::global_store::with_store;
use crate::t::{self, ConcreteTypeBuilder, TypeBuilder};
use crate::{errors::Result, types::TypeId};

use super::{where_::Where, with_filters::WithFilters, TypeGen};

pub struct QueryWhereExpr {
    model_id: TypeId,
}

impl QueryWhereExpr {
    pub fn new(model_id: TypeId) -> Self {
        Self { model_id }
    }
}

impl TypeGen for QueryWhereExpr {
    fn generate(&self, context: &mut super::TypeGenContext) -> Result<TypeId> {
        let where_type = context.generate(&Where::new(self.model_id, true))?;
        let extended_type =
            context.generate(&WithFilters::new(where_type, self.model_id, false))?;

        let props = with_store(|s| extended_type.as_struct(s).unwrap().data.props.to_vec());

        let name = self.name(context);
        let self_ref = t::proxy(&name).build()?;
        let and = t::optional(t::array(self_ref).build()?).build()?;

        let mut builder = t::struct_();
        builder.named(name);
        for (k, ty) in props.into_iter() {
            builder.prop(k.clone(), ty.into());
        }
        builder.prop("AND", and);
        builder.prop("OR", and);
        builder.prop("NOT", t::optional(self_ref).build()?);

        builder.build()
    }

    fn name(&self, context: &super::TypeGenContext) -> String {
        format!(
            "Query{}WhereInput",
            context.registry.models.get(&self.model_id).unwrap().name
        )
    }
}
