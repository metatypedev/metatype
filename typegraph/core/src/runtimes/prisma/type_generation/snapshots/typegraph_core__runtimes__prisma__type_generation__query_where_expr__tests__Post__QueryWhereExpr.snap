---
source: typegraph/core/src/runtimes/prisma/type_generation/query_where_expr.rs
expression: "tree::print(post_where_expr)"
---
root: struct 'QueryPostWhereInput' #78
├─ [id]: optional '_58__integer_filter_c?' #58
│  └─ item: optional '_integer_filter_c' #20
│     └─ item: union #19
│        ├─ variant_0: either '_integer_filter' #17
│        │  ├─ variant_0: integer #9
│        │  ├─ variant_1: struct #12
│        │  │  └─ [equals]: integer #9
│        │  ├─ variant_2: struct #13
│        │  │  └─ [not]: integer #9
│        │  ├─ variant_3: struct #14
│        │  │  ├─ [lt]: optional #10
│        │  │  │  └─ item: integer #9
│        │  │  ├─ [gt]: optional #10
│        │  │  │  └─ item: integer #9
│        │  │  ├─ [lte]: optional #10
│        │  │  │  └─ item: integer #9
│        │  │  └─ [gte]: optional #10
│        │  │     └─ item: integer #9
│        │  ├─ variant_4: struct #15
│        │  │  └─ [in]: array #11
│        │  │     └─ item: integer #9
│        │  └─ variant_5: struct #16
│        │     └─ [notIn]: array #11
│        │        └─ item: integer #9
│        └─ variant_1: struct #18
│           └─ [not]: either '_integer_filter' #17
│              ├─ variant_0: integer #9
│              ├─ variant_1: struct #12
│              │  └─ [equals]: integer #9
│              ├─ variant_2: struct #13
│              │  └─ [not]: integer #9
│              ├─ variant_3: struct #14
│              │  ├─ [lt]: optional #10
│              │  │  └─ item: integer #9
│              │  ├─ [gt]: optional #10
│              │  │  └─ item: integer #9
│              │  ├─ [lte]: optional #10
│              │  │  └─ item: integer #9
│              │  └─ [gte]: optional #10
│              │     └─ item: integer #9
│              ├─ variant_4: struct #15
│              │  └─ [in]: array #11
│              │     └─ item: integer #9
│              └─ variant_5: struct #16
│                 └─ [notIn]: array #11
│                    └─ item: integer #9
├─ [title]: optional '_59__string_filter_c?' #59
│  └─ item: optional '_string_filter_c' #37
│     └─ item: union #36
│        ├─ variant_0: union '_string_filter' #34
│        │  ├─ variant_0: string #22
│        │  ├─ variant_1: struct #25
│        │  │  └─ [equals]: string #22
│        │  ├─ variant_2: struct #26
│        │  │  └─ [not]: string #22
│        │  ├─ variant_3: struct #27
│        │  │  └─ [in]: array #24
│        │  │     └─ item: string #22
│        │  ├─ variant_4: struct #28
│        │  │  └─ [notIn]: array #24
│        │  │     └─ item: string #22
│        │  ├─ variant_5: struct #31
│        │  │  ├─ [contains]: string #22
│        │  │  └─ [mode]: optional #30
│        │  │     └─ item: string #29 enum{ '"insensitive"' }
│        │  ├─ variant_6: struct #32
│        │  │  └─ [search]: string #22
│        │  └─ variant_7: struct #33
│        │     ├─ [startsWith]: optional #23
│        │     │  └─ item: string #22
│        │     └─ [endsWith]: optional #23
│        │        └─ item: string #22
│        └─ variant_1: struct #35
│           └─ [not]: union '_string_filter' #34
│              ├─ variant_0: string #22
│              ├─ variant_1: struct #25
│              │  └─ [equals]: string #22
│              ├─ variant_2: struct #26
│              │  └─ [not]: string #22
│              ├─ variant_3: struct #27
│              │  └─ [in]: array #24
│              │     └─ item: string #22
│              ├─ variant_4: struct #28
│              │  └─ [notIn]: array #24
│              │     └─ item: string #22
│              ├─ variant_5: struct #31
│              │  ├─ [contains]: string #22
│              │  └─ [mode]: optional #30
│              │     └─ item: string #29 enum{ '"insensitive"' }
│              ├─ variant_6: struct #32
│              │  └─ [search]: string #22
│              └─ variant_7: struct #33
│                 ├─ [startsWith]: optional #23
│                 │  └─ item: string #22
│                 └─ [endsWith]: optional #23
│                    └─ item: string #22
├─ [author]: optional '_72_UserWhere__skip_8_PostWhere?' #72
│  └─ item: struct 'UserWhere__skip_8_PostWhere' #71
│     ├─ [id]: optional '_60__integer_filter_c?' #60
│     │  └─ item: optional '_integer_filter_c' #20
│     │     └─ item: union #19
│     │        ├─ variant_0: either '_integer_filter' #17
│     │        │  ├─ variant_0: integer #9
│     │        │  ├─ variant_1: struct #12
│     │        │  │  └─ [equals]: integer #9
│     │        │  ├─ variant_2: struct #13
│     │        │  │  └─ [not]: integer #9
│     │        │  ├─ variant_3: struct #14
│     │        │  │  ├─ [lt]: optional #10
│     │        │  │  │  └─ item: integer #9
│     │        │  │  ├─ [gt]: optional #10
│     │        │  │  │  └─ item: integer #9
│     │        │  │  ├─ [lte]: optional #10
│     │        │  │  │  └─ item: integer #9
│     │        │  │  └─ [gte]: optional #10
│     │        │  │     └─ item: integer #9
│     │        │  ├─ variant_4: struct #15
│     │        │  │  └─ [in]: array #11
│     │        │  │     └─ item: integer #9
│     │        │  └─ variant_5: struct #16
│     │        │     └─ [notIn]: array #11
│     │        │        └─ item: integer #9
│     │        └─ variant_1: struct #18
│     │           └─ [not]: either '_integer_filter' #17
│     │              ├─ variant_0: integer #9
│     │              ├─ variant_1: struct #12
│     │              │  └─ [equals]: integer #9
│     │              ├─ variant_2: struct #13
│     │              │  └─ [not]: integer #9
│     │              ├─ variant_3: struct #14
│     │              │  ├─ [lt]: optional #10
│     │              │  │  └─ item: integer #9
│     │              │  ├─ [gt]: optional #10
│     │              │  │  └─ item: integer #9
│     │              │  ├─ [lte]: optional #10
│     │              │  │  └─ item: integer #9
│     │              │  └─ [gte]: optional #10
│     │              │     └─ item: integer #9
│     │              ├─ variant_4: struct #15
│     │              │  └─ [in]: array #11
│     │              │     └─ item: integer #9
│     │              └─ variant_5: struct #16
│     │                 └─ [notIn]: array #11
│     │                    └─ item: integer #9
│     ├─ [name]: optional '_61__string_filter_c?' #61
│     │  └─ item: optional '_string_filter_c' #37
│     │     └─ item: union #36
│     │        ├─ variant_0: union '_string_filter' #34
│     │        │  ├─ variant_0: string #22
│     │        │  ├─ variant_1: struct #25
│     │        │  │  └─ [equals]: string #22
│     │        │  ├─ variant_2: struct #26
│     │        │  │  └─ [not]: string #22
│     │        │  ├─ variant_3: struct #27
│     │        │  │  └─ [in]: array #24
│     │        │  │     └─ item: string #22
│     │        │  ├─ variant_4: struct #28
│     │        │  │  └─ [notIn]: array #24
│     │        │  │     └─ item: string #22
│     │        │  ├─ variant_5: struct #31
│     │        │  │  ├─ [contains]: string #22
│     │        │  │  └─ [mode]: optional #30
│     │        │  │     └─ item: string #29 enum{ '"insensitive"' }
│     │        │  ├─ variant_6: struct #32
│     │        │  │  └─ [search]: string #22
│     │        │  └─ variant_7: struct #33
│     │        │     ├─ [startsWith]: optional #23
│     │        │     │  └─ item: string #22
│     │        │     └─ [endsWith]: optional #23
│     │        │        └─ item: string #22
│     │        └─ variant_1: struct #35
│     │           └─ [not]: union '_string_filter' #34
│     │              ├─ variant_0: string #22
│     │              ├─ variant_1: struct #25
│     │              │  └─ [equals]: string #22
│     │              ├─ variant_2: struct #26
│     │              │  └─ [not]: string #22
│     │              ├─ variant_3: struct #27
│     │              │  └─ [in]: array #24
│     │              │     └─ item: string #22
│     │              ├─ variant_4: struct #28
│     │              │  └─ [notIn]: array #24
│     │              │     └─ item: string #22
│     │              ├─ variant_5: struct #31
│     │              │  ├─ [contains]: string #22
│     │              │  └─ [mode]: optional #30
│     │              │     └─ item: string #29 enum{ '"insensitive"' }
│     │              ├─ variant_6: struct #32
│     │              │  └─ [search]: string #22
│     │              └─ variant_7: struct #33
│     │                 ├─ [startsWith]: optional #23
│     │                 │  └─ item: string #22
│     │                 └─ [endsWith]: optional #23
│     │                    └─ item: string #22
│     └─ [posts]: optional #70
│        └─ item: union #69
│           ├─ variant_0: struct #64
│           │  └─ [every]: optional '_63_PostWhere?' #63
│           │     └─ item: &PostWhere #62
│           ├─ variant_1: struct #66
│           │  └─ [some]: optional '_65_PostWhere?' #65
│           │     └─ item: &PostWhere #62
│           └─ variant_2: struct #68
│              └─ [none]: optional '_67_PostWhere?' #67
│                 └─ item: &PostWhere #62
├─ [AND]: optional '_76__75_QueryPostWhereInput[]?' #76
│  └─ item: array '_75_QueryPostWhereInput[]' #75
│     └─ item: &QueryPostWhereInput #74
├─ [OR]: optional '_76__75_QueryPostWhereInput[]?' #76
│  └─ item: array '_75_QueryPostWhereInput[]' #75
│     └─ item: &QueryPostWhereInput #74
└─ [NOT]: optional '_77_QueryPostWhereInput?' #77
   └─ item: &QueryPostWhereInput #74

